py-rse/py-scripting.Rmd

```diff
@@ -659,69 +837,63 @@ def main(args):
                         help='Limit output to n most frequent words')
     args = parser.parse_args()
     main(args)
 ```
 
-> **Module naming conventions**
+Any Python source file can be imported by any other.
+This is why Python files should be named using [`snake_case`][snake-case]
+instead of [`kebab-case`][kebab-case]:
+an expression like `import some-thing` isn't allowed
+because `some-thing` isn't a legal variable name.
+When a file is imported,
+the statements in it are executed as it loads.
+Variables, functions, and items defined in the file are then available as `module.thing`,
+where `module` is the filename (without the `.py` extension)
+and `thing` is the name of the item.
+
+> **The `__pycache__` Directory**
 >
-> Any Python source file can be imported by any other.
-> This is why Python files should be named using [`snake_case`][snake-case]
-> instead of [`kebab-case`][kebab-case]
-> (an expression like `import some-thing` isn't allowed
-> because `some-thing` isn't a legal variable name.)
-> When a file is imported,
-> the statements in it are executed as it loads.
-> Variables, functions, and items defined in the file are
-> then available as `module.name`, where `module` is the
-> filename (without the `.py` extension) and `name` is the
-> name of the item.
-
-> **Ignoring the `__pycache__`**
+> When we import a file,
+> Python translates the source code into instructions called [byte codes][byte-code]
+> that it can execute efficiently.
+> Since the byte codes only change when the source changes,
+> Python saves the byte code in a separate file,
+> and reloads that file instead of re-translating the source code
+> the next time it's asked to import the file
+> (unless the file has changed,
+> in which case Python starts from the beginning).
 >
-> It also creates a subdirectory called `__pycache__`
-> that holds the compiled versions of the imported files.
-> The next time Python imports the module,
-> it checks the timestamp on the script
-> and the timestamp on the corresponding file in `__pycache__`.
-> If the latter is more recent,
-> Python doesn't bother to recompile the file:
-> it just loads the bytes in the cached version and uses those.
-> In the general spirit of ["Don't Repeat Yourself (DRY)"][dry]
-> we typically don't put both the original and processed versions of files in
-> version control,
-> so we normally add the `__pycache__` directory to in `.gitignore`
-> (under some circumstances,
-> it can be convenient to have `__pychache__` under version control,
-> but then you have to have processes in place
-> to make sure that they stay in sync).
-
-## Mission Critical Exercise {#py-rse-py-scripting-critical-exercise}
+> Python creates a subdirectory called `__pycache__`
+> that holds the byte code for the files imported from that directory.
+> We typically don't want to put the files in `__pycache__` in version control,
+> so we normally add `__pycache__` to `.gitignore`.
```

Comment:
```suggestion
> so we normally add `__pycache__` to `.gitignore` (Section \@ref(py-rse-git-cmdline-ignore)).
```

Reply:
Because we haven't met version control yet or `.gitignore` yet...

Reply:
Reworded to avoid mentioning `.gitignore` but include the forward reference.